### PR TITLE
Allow configuring where artifacts are downloaded from

### PR DESCRIPTION
**Pull request by emilyalbini**
Bootstrap has support for downloading prebuilt LLVM and rustc artifacts to speed up local builds, but that currently works only for users working on `rust-lang/rust`. Forks of the repository (for example Ferrocene) might have different URLs to download artifacts from, or might use a different email address on merge commits, breaking both LLVM and rustc artifact downloads.

This PR refactors bootstrap to load the download URLs and other constants from `src/stage0.json`, allowing downstream forks to tweak those values. It also future-proofs the download code to easily allow forks to add their own custom protocols (like `s3://`).

This PR is best reviewed commit-by-commit.

### CODE REVIEW

File: src/bootstrap/builder.rs
Comment (bjorn3) [inline src/bootstrap/builder.rs:878]: Maybe move this match into `download_with_retries`?

Comment (pietroalbini) [reply src/bootstrap/builder.rs:878]: Hmm, is the move to `download_with_retries` just for code organization, or because you'd like retries to be handled regardless of the protocol?

In the latter case, that'd require a larger refactor of `download_with_retries`:

* Curl retries are not handled in Rust code, but through the `--retries` flag. We'd need to move that into Rust code.
* HTTP first tries curl, and then on Windows it tries PowerShell. Encoding that in a generic retry mechanism would be slightly annoying.

File: src/bootstrap/config.rs
Comment (jyn514) [inline src/bootstrap/config.rs:1399]: My rustfmt commit also introduced a stage0 struct - do you mind waiting for that to merge sowe don't have duplicates? I'll try and get that done today.

**Review by jyn514 — COMMENTED** [review]
A lot of this code is being changed significantly in my rustfmt PR (https://github.com/rust-lang/rust/pull/97507) - I'd prefer to wait until that's merged.

File: src/bootstrap/config.rs
Comment (pietroalbini) [reply src/bootstrap/config.rs:1399]: Sure! I'll wait and then rebase :sligh

File: src/bootstrap/builder.rs
Comment (bjorn3) [reply src/bootstrap/builder.rs:878]: I assumed `download_with_retries` would be called from other places, but it only has a single callsite. Keeping the match where it currently is is fine.

**Comment by pietroalbini** [comment]
Rebased on top of master!

r? @jyn514

File: src/stage0.json
Comment (jyn514) [inline src/stage0.json:2]: nit: why nest this under a `config` key?

**Review by jyn514 — APPROVED** [review]
Changes look good to me - I assume you've already tested this PR with your fork and made sure you can download the artifacts by changing only stage0.json? Testing that there aren't merge conflicts seems like overkill, but you'll have to edit stage0.json after this is merged anyway.

File: src/stage0.json
Comment (pietroalbini) [reply src/stage0.json:2]: Because that way the logic in `bump-stage0` to preserve the contents of these keys doesn't need to be updated as new keys are added.

**Comment by pietroalbini** [comment]
> I assume you've already tested this PR with your fork and made sure you can download the artifacts by changing only stage0.json?

I can double-check on Monday (tested an old iteration of this PR).

**Comment by pietroalbini** [comment]
Just confirmed that yes, this works! Also pushed a small commit to move `Stage0` closer to `Config` (it's easier to see where it is, and it avoids a merge conflict on the Ferrocene side).

**Comment by jyn514** [comment]
Missed that this was waiting on me, sorry - this lgtm :)

@bors r+

**Pull request merged**
